tests/core/snapd-failover/task.yaml

```diff
@@ -2,61 +2,123 @@ summary: Check that snapd failure handling works
+
+        echo "Verify that a random signal does not trigger the failure handling"
+        echo "and snapd is just restarted"
+        systemctl kill --signal=SIGSEGV snapd.service
+
+        started_after_rand_sig="$(get_journalctl_log -u snapd.failure | grep -c 'Started Failure handling of the snapd snap.' || true)"
+
+        if [ "$started_after_rand_sig" -gt "$started_before" ] ; then
+            echo "snapd.failure inadvertently started with a random signal!"
+            exit 1
+        fi
+
+        echo "Snap list is working still"
+        snap list | MATCH "^snapd .* $current .*"
+
+        SNAPD_SNAP=$(ls /var/lib/snapd/snaps/snapd_"$current".snap)
+    else
+        # this will be used for uc16 tests a bit later
+        current=$(readlink /snap/core/current)
+
+        snap download snapd --basename=snapd
+        SNAPD_SNAP=$(pwd)/snapd.snap
+    fi
+
 
     echo "Break snapd"
     unsquashfs -d ./snapd-broken "$SNAPD_SNAP"
     echo "" > ./snapd-broken/usr/lib/snapd/snapd
+
+    if is_core16_system; then
+        # pick up snap-failure from the current core snap as that's where the 
+        # updated code under test is located
+        cp -a /usr/lib/snapd/snap-failure ./snapd-broken/usr/lib/snapd/snap-failure
+    fi
+
     (cd ./snapd-broken && snap pack .)
     echo "Now install the broken snapd"
     if snap install --dangerous ./snapd-broken/snapd_*.snap; then
         echo "installing a broken snapd should not work, test broken"
         exit 1
     fi
 
+    echo "Verify that snapd.failure was activated when we tried to install a broken snapd"
+    for _ in $(seq 60); do
+        # get the number of times that snapd.failure was started
+        started_after="$(get_journalctl_log -u snapd.failure | grep -c 'Started Failure handling of the snapd snap.' || true)"
+        if [ "$started_after" -gt "$started_before" ] ; then
+            break
+        fi
+        sleep 1
+    done
+
+    # just because snapd.failure.service is active doesn't mean that we are 
+    # fully ready; we should wait until the snap command shows up again
     echo "And verify that snap commands still work and snapd is reverted"
-    retry-tool -n 30 --wait 5 systemctl is-active snapd.failure.service
-    snap list | MATCH "^snapd .* $current .*"
+    echo "Verify that we were rolled back to the previous snapd"
+    if is_core16_system; then
+        retry-tool -n 30 --wait 1 bash -c 'snap list | MATCH "^core .* $current .*"'
+    else
+        retry-tool -n 30 --wait 1 bash -c 'snap list | MATCH "^snapd .* $current .*"'
+    fi
 
     echo "Verify we got the expected error message"
     snap change --last=install|MATCH "there was a snapd rollback across the restart"
 
     echo "restart snapd and ensure we can still talk to it"
     systemctl restart snapd.socket snapd.service
-    snap list | MATCH "^snapd .* $current .*"
+
+    if is_core16_system; then
+        # on UC16 we should have rolled back to the core snap, and there should
```

Comment:
There's 2 possible scenarios on UC16. First one is like in `tests/main/ubuntu-core-snapd-failover/task.yaml`, which is a failover on first install of snapd snap (got merged here). The second one is like UC18, where there's already a working revision of the snapd snap installed and a new 'broken' one gets added. In this scenario, the failover reverts to the previous revision of the snapd snap. This should have been enabled in this test after the PR allowing snapd on core landed, but didn't. Maybe it's a good time to add it now. I think I'd keep `tests/main/ubuntu-core-snapd-failover/task.yaml`, possibly renaming it to `snapd-failover-on-first-install`?

Reply:
Hmm, I suppose in this case, we could have the snapd-failover test for UC16 be in the situation where the snapd snap is already installed on UC16 so that all the logic here is the same for all systems, there's just a UC16 specific prepare section to do the transition. 

Then we would have a separate test like you had originally which is similar to snapd-failover, but just about the transition from core -> snapd. 

WDYT? 

Note that I do want to keep all ubuntu core specific tests in the "tests/core" suite as that's the new place for all UC specific tests and only is setup to run with ubuntu-core-* systems.

Reply:
Yeah, this sounds fine. It's probably easiest to restore `tests/main/ubuntu-core-snapd-failover/task.yaml` and move it under `tests/core` and rename to something more meaningful.

Reply:
I ended up renaming it to `tests/core/core-to-snapd-failover16`, let me know if you have strong feelings about the test name